packages/jsii/lib/transforms/deprecated-warnings-transformer.ts

```diff
@@ -0,0 +1,354 @@
+      );
+
+      if (isDeprecated(node)) {
+        this.warnings.push(
+          getWarning(this.typeChecker, node, ElementType.METHOD),
+        );
+      }
+
+      this.addWarning = this.warnings.length > 0;
+    }
+
+    if (ts.isBlock(node) && this.addWarning) {
+      const block = node as Block;
+      const nodeArray: ts.NodeArray<ts.Statement> = ts.createNodeArray([
+        ...this.createWarningStatements(),
+        ...block.statements,
+      ]);
+
+      // "Unstack" all warnings specific to this block
+      this.warnings = this.warnings.filter(
+        (warning) =>
+          ![ElementType.METHOD, ElementType.PARAMETER].includes(warning.type),
+      );
+      return ts.updateBlock(node, nodeArray) as any;
+    }
+
+    return this.visitEachChild(node);
+  }
+
+  private getParameterWarnings(node: MethodLikeDeclaration): Warning[] {
+    return node.parameters
+      .filter((parameter) =>
+        this.typesWithDeprecatedProperties.has(
+          this.typeChecker.getTypeAtLocation(parameter),
+        ),
+      )
+      .map(
+        (parameter) =>
+          getWarning(
+            this.typeChecker,
+            this.typesWithDeprecatedProperties.get(
+              this.typeChecker.getTypeAtLocation(parameter),
+            )!,
+            ElementType.PARAMETER,
+          )!,
+      );
+  }
+
+  private createWarningStatements(): ts.Statement[] {
+    return [...this.warnings].flatMap(createWarningStatement);
+  }
+
+  // TODO Rename this method
+  private findDeprecatedInTheTypeHierarchy(
+    toProcess: ts.Type[],
+  ): ts.Type | undefined {
+    if (toProcess.length === 0) {
+      return undefined;
+    }
+
+    if (this.deprecatedTypes.includes(toProcess[0])) {
+      return toProcess[0];
+    }
+
+    const node = this.index.get(
+      this.typeChecker.getFullyQualifiedName(toProcess[0].symbol),
+    )!;
+
+    // This type was not declared in the JSII assembly. Skipping.
+    if (node == null) {
+      return undefined;
+    }
+
+    if (!ts.isClassDeclaration(node) && !ts.isInterfaceDeclaration(node)) {
+      throw new Error('Node should be a class or interface declaration.');
+    }
+
+    const newBatch = toProcess.slice(1);
+    const heritageClauses = node.heritageClauses ?? [];
+    for (const clause of heritageClauses) {
+      for (const expression of clause.types) {
+        const type = this.typeChecker.getTypeAtLocation(expression);
+        if (!toProcess.includes(type)) {
+          newBatch.push(type);
+        }
+      }
+    }
+
+    return this.findDeprecatedInTheTypeHierarchy(newBatch);
+  }
+}
+
+function createWarningStatement(warning: Warning): ts.Statement[] {
+  const message = `${warning?.elementName} is deprecated.\n  ${warning?.message}\n  This API will be removed in the net major release.`;
+  // TODO Are random numbers a good idea?
+  const functionName = `printWarning${getRandomInt(
+    0,
+    Number.MAX_SAFE_INTEGER,
+  )}`;
+
+  // TODO There must be a simpler way...
```

Comment:
I tried using `ts.createSourceFile()` with the whole code, but it makes a mess in the generated file.